specific/cmo_sale/models/sale.py

```diff
@@ -93,11 +141,41 @@ def _compute_project_number(self):
 
     @api.model
     def _default_covenant(self):
-        convenants = self.env['sale.convenant.description'].search([
-            ['active', '=', True],
-        ])
-        if convenants:
-            return convenants[0].description
+        Description = self.env['sale.covenant.description']
+        covenants = Description.search([('active', '=', True), ])
+        return covenants and covenants[0].description or False
+
+    @api.multi
+    @api.constrains('order_line')
+    def _constrains_order_line(self):
+        self.ensure_one()
+        if not self.order_line:
+            raise ValidationError("Must have at least 1 order line!")
+        else:
+            for line in self.order_line:
+                if ((line.price_unit <= 0) or (line.product_uom_qty <= 0)) and \
+                   (line.order_lines_group == 'before'):
+                    raise ValidationError(
+                        "Unit Price and Quantity in order \
+                        line must more than zero !"
+                    )
+
+    @api.multi
+    def _get_amount_by_custom_group(self, custom_group):
+        self.ensure_one()
+        lines = self.order_line.filtered(
+            lambda r:
+            (r.order_lines_group == 'before') and
+            (r.sale_layout_custom_group_id.id == custom_group.id)
+        )
+        return sum(lines.mapped('price_subtotal'))
+
+    @api.multi
+    def _compute_margin_percentage(self):
+        for order in self:
+            if order.amount_untaxed != 0:
```

Comment:
amount_untaxed is Float. So, we should compare with 0.0
Note: no harm, but it is nicer.